File: java/Apparat.Core/src/com/joa_ebert/apparat/taas/toolkit/generic/CopyPropagation.java
/*
 * This file is part of Apparat.
 * 
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 * 
 * Copyright (C) 2009 Joa Ebert
 * http://www.joa-ebert.com/
 * 
 */

package com.joa_ebert.apparat.taas.toolkit.generic;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;
import java.util.logging.Logger;

import com.joa_ebert.apparat.abc.AbcEnvironment;
import com.joa_ebert.apparat.controlflow.BasicBlock;
import com.joa_ebert.apparat.controlflow.BasicBlockGraph;
import com.joa_ebert.apparat.controlflow.ControlFlowGraphException;
import com.joa_ebert.apparat.controlflow.VertexKind;
import com.joa_ebert.apparat.taas.TaasConstant;
import com.joa_ebert.apparat.taas.TaasException;
import com.joa_ebert.apparat.taas.TaasLocal;
import com.joa_ebert.apparat.taas.TaasMethod;
import com.joa_ebert.apparat.taas.TaasValue;
import com.joa_ebert.apparat.taas.TaasVertex;
import com.joa_ebert.apparat.taas.compiler.TaasCompiler;
import com.joa_ebert.apparat.taas.expr.AbstractLocalExpr;
import com.joa_ebert.apparat.taas.expr.TSetLocal;
import com.joa_ebert.apparat.taas.toolkit.ITaasTool;
import com.joa_ebert.apparat.taas.toolkit.TaasToolkit;
import com.joa_ebert.apparat.taas.toolkit.livenessAnalysis.LivenessAnalysis;

/**
 * 
 * @author Joa Ebert
 * 
 */
public final class CopyPropagation implements ITaasTool
{
	private static final boolean DEBUG = false;

	private static final Logger LOG = ( DEBUG ) ? Logger
			.getLogger( CopyPropagation.class.getName() ) : null;

	private boolean cp( final AbcEnvironment environment,
			final TaasMethod method ) throws ControlFlowGraphException
	{
		if( DEBUG )
		{
			LOG.entering( "CopyPropagation", "cp" );
			LOG.info( "Begin CopyPropagation..." );
		}

		boolean changed = false;

		final LivenessAnalysis la = new LivenessAnalysis( method );
		final BasicBlockGraph<TaasVertex> graph = la.getGraph();

		la.solve();

		for( final BasicBlock<TaasVertex> block : graph.vertexList() )
		{
			final List<TaasLocal> liveOut = la.liveOut( block );
			final List<TaasVertex> vertices = block.vertices();

			final List<TSetLocal> locals = new LinkedList<TSetLocal>();

			for( final TaasVertex vertex : vertices )
			{
				if( VertexKind.Default != vertex.kind )
				{
					continue;
				}

				final TaasValue value = vertex.value;

				if( value instanceof TSetLocal )
				{
					final TSetLocal setLocal = (TSetLocal)value;

					if( !setLocal.hasSideEffects()
							&& !liveOut.contains( setLocal.local ) )
					{
						if( DEBUG )
						{
							LOG.info( "Candidate: " + setLocal );
						}

						locals.add( setLocal );
					}
				}
			}

			nextLocal: for( final TSetLocal setLocal : locals )
			{
				if( DEBUG )
				{
					LOG.info( "Processing " + setLocal );
				}

				if( setLocal.value instanceof TaasConstant )
				{
					if( DEBUG )
					{
						LOG.info( "Constant assignment: x = const" );
					}

					//
					// Constant assignment:
					//
					// x = const
					//

					final ListIterator<TaasVertex> iter = vertices
							.listIterator();

					final List<TaasVertex> verticesToRemove = new LinkedList<TaasVertex>();

					boolean startPropagation = false;

					while( iter.hasNext() )
					{
						final TaasVertex vertex = iter.next();

						if( VertexKind.Default != vertex.kind )
						{
							continue;
						}

						final TaasValue value = vertex.value;

						if( value == setLocal )
						{
							if( DEBUG )
							{
								LOG.info( "Start propagation at " + value );
							}

							startPropagation = true;
							verticesToRemove.add( vertex );
							continue;
						}
						else if( value instanceof AbstractLocalExpr
								&& ( (AbstractLocalExpr)value ).local
										.getIndex() == setLocal.local
										.getIndex() )
						{
							if( DEBUG )
							{
								LOG.info( "Stop propagation at " + value );
							}

							continue nextLocal;
						}
						else if( TaasToolkit.references( value, setLocal.local ) )
						{
							if( startPropagation )
							{
								if( DEBUG )
								{
									LOG.info( "Propagating copy into " + value );
								}

								TaasToolkit.replace( value, setLocal.local,
										setLocal.value.dup() );

								changed = true;
							}
						}
					}

					for( final TaasVertex vertex : verticesToRemove )
					{
						if( DEBUG )
						{
							LOG.info( "Removing vertex " + vertex );
						}

						TaasToolkit.remove( method, vertex );
						changed = true;
					}
				}
				else if( setLocal.value instanceof TaasLocal )
				{
					if( DEBUG )
					{
						LOG.info( "Copy assignment: x = y" );
					}

					//
					// Copy assignment:
					//
					// x = y;
					//

					final TaasLocal copy = (TaasLocal)setLocal.value;
					final ListIterator<TaasVertex> iter = vertices
							.listIterator();

					if( DEBUG )
					{
						LOG.info( "Copy is " + copy );
					}

					final List<TaasVertex> verticesToRemove = new LinkedList<TaasVertex>();

					boolean startPropagation = false;

					while( iter.hasNext() )
					{
						final TaasVertex vertex = iter.next();

						if( VertexKind.Default != vertex.kind )
						{
							continue;
						}

						final TaasValue value = vertex.value;

						if( value == setLocal )
						{
							if( DEBUG )
							{
								LOG.info( "Start propagation at " + value );
							}

							startPropagation = true;
							verticesToRemove.add( vertex );
							continue;
						}
						else if( value instanceof AbstractLocalExpr
								&& ( ( (AbstractLocalExpr)value ).local
										.getIndex() == setLocal.local
										.getIndex() || ( (AbstractLocalExpr)value ).local
										.getIndex() == copy.getIndex() ) )
						{
							//
							// || ( (AbstractLocalExpr)value ).local
							// .getIndex() == copy.getIndex() )
							//

							if( DEBUG )
							{
								LOG.info( "Stop propagation at " + value );
							}

							continue nextLocal;
						}
						else if( TaasToolkit.references( value, setLocal.local ) )
						{
							if( startPropagation )
							{
								if( DEBUG )
								{
									LOG.info( "Propagating copy into " + value );
								}

								TaasToolkit.replace( value, setLocal.local,
										setLocal.value );

								changed = true;
							}
						}
					}

					for( final TaasVertex vertex : verticesToRemove )
					{
						if( DEBUG )
						{
							LOG.info( "Removing vertex " + vertex );
						}

						TaasToolkit.remove( method, vertex );
						changed = true;
					}
				}
				else
				{
					if( DEBUG )
					{
						LOG.info( "Other assignment: x = expr" );
					}

					//
					// Other assignment:
					//
					// x = 2 * y;
					// x = Math.random();
					// .
					// .
					// .
					// etc.
					//

					ListIterator<TaasVertex> iter = vertices.listIterator();

					int uses = 0;

					boolean startPropagation = false;

					while( iter.hasNext() )
					{
						final TaasVertex vertex = iter.next();

						if( VertexKind.Default != vertex.kind )
						{
							continue;
						}

						final TaasValue value = vertex.value;

						if( value == setLocal )
						{
							startPropagation = true;
							continue;
						}
						else if( value instanceof AbstractLocalExpr
								&& ( (AbstractLocalExpr)value ).local
										.getIndex() == setLocal.local
										.getIndex() )
						{
							continue nextLocal;
						}
						else
						{
							if( startPropagation )
							{
								uses += TaasToolkit.numReferences( value,
										setLocal.local );
							}
						}
					}

					if( DEBUG )
					{
						LOG.info( "Value uses: " + uses );
					}

					if( 1 == uses )
					{
						startPropagation = false;

						iter = vertices.listIterator();

						final List<TaasVertex> verticesToRemove = new LinkedList<TaasVertex>();

						final LinkedList<TaasLocal> sideEffects = new LinkedList<TaasLocal>();

						TaasToolkit.searchAll( setLocal, TaasLocal.class, true,
								sideEffects );

						while( iter.hasNext() )
						{
							final TaasVertex vertex = iter.next();

							if( VertexKind.Default != vertex.kind )
							{
								continue;
							}

							final TaasValue value = vertex.value;

							if( value == setLocal )
							{
								if( DEBUG )
								{
									LOG.info( "Start propagation at " + value );
								}

								startPropagation = true;
								verticesToRemove.add( vertex );
								continue;
							}
							else if( value instanceof AbstractLocalExpr )
							{
								if( ( (AbstractLocalExpr)value ).local
										.getIndex() == setLocal.local
										.getIndex() )
								{
									if( DEBUG )
									{
										LOG.info( "Stop propagation at "
												+ value );
									}

									continue nextLocal;
								}

								if( value instanceof TSetLocal )
								{
									final TSetLocal setLocal2 = (TSetLocal)value;

									for( final TaasLocal local : sideEffects )
									{
										if( setLocal2.local.getIndex() == local
												.getIndex() )
										{
											if( startPropagation
													&& TaasToolkit.references(
															value,
															setLocal.local ) )
											{
												if( DEBUG )
												{
													LOG
															.info( "Propagating value into "
																	+ value );
												}

												TaasToolkit.replace( value,
														setLocal.local,
														setLocal.value );

												changed = true;
											}

											if( startPropagation )
											{
												if( DEBUG )
												{
													LOG
															.info( "Stop propagation at "
																	+ value );
												}

												continue nextLocal;
											}
										}
									}

									if( startPropagation
											&& TaasToolkit.references( value,
													setLocal.local ) )
									{
										if( DEBUG )
										{
											LOG.info( "Propagating value into "
													+ value );
										}

										TaasToolkit.replace( value,
												setLocal.local, setLocal.value );

										changed = true;
									}
								}
								else if( TaasToolkit.references( value,
										setLocal.local ) )
								{
									if( startPropagation )
									{
										if( DEBUG )
										{
											LOG.info( "Propagating value into "
													+ value );
										}

										TaasToolkit.replace( value,
												setLocal.local, setLocal.value );

										changed = true;
									}
								}
							}
							else if( TaasToolkit.references( value,
									setLocal.local ) )
							{
								if( startPropagation )
								{
									if( DEBUG )
									{
										LOG.info( "Propagating value into "
												+ value );
									}

									TaasToolkit.replace( value, setLocal.local,
											setLocal.value );

									changed = true;
								}
							}
						}

						for( final TaasVertex vertex : verticesToRemove )
						{
							if( DEBUG )
							{
								LOG.info( "Removing vertex " + vertex );
							}

							TaasToolkit.remove( method, vertex );
							changed = true;
						}
					}
				}
			}
		}

		if( DEBUG )
		{
			LOG.info( "Changed: " + changed );
			LOG.exiting( "CopyPropagation", "cp" );
		}

		return changed;
	}

	public boolean manipulate( final AbcEnvironment environment,
			final TaasMethod method )
	{
		boolean changed = false;

		try
		{
			changed = cp( environment, method );
		}
		catch( final ControlFlowGraphException exception )
		{
			throw new TaasException( exception );
		}

		try
		{
			final LinkedList<TaasVertex> list = method.code.vertexList();
			final LinkedList<TaasVertex> removes = new LinkedList<TaasVertex>();
			final Iterator<TaasVertex> iter = list.listIterator();

			while( iter.hasNext() )
			{
				final TaasVertex vertex = iter.next();

				if( VertexKind.Default != vertex.kind )
				{
					continue;
				}

				final TaasValue value = vertex.value;

				if( value instanceof TaasConstant && value.isConstant() )
				{
					removes.add( vertex );
				}
				else if( value instanceof TaasLocal )
				{
					removes.add( vertex );
				}
				else
				{
					final Iterator<TaasVertex> refIter = list
							.descendingIterator();

					boolean canRemove = false;

					while( refIter.hasNext() )
					{
						final TaasVertex refVert = refIter.next();
						final TaasValue refValue = refVert.value;

						if( refValue == value )
						{
							break;
						}
						else if( TaasToolkit.references( refValue, value ) )
						{
							canRemove = true;
							break;
						}
					}

					if( canRemove )
					{
						removes.add( vertex );
					}
				}
			}

			if( !changed )
			{
				changed = !removes.isEmpty();
			}

			for( final TaasVertex vertex : removes )
			{
				TaasToolkit.remove( method, vertex );
			}
		}
		catch( final ControlFlowGraphException e )
		{
			throw new TaasException( e );
		}

		if( TaasCompiler.SHOW_ALL_TRANSFORMATIONS && changed )
		{
			TaasToolkit.debug( "CopyPropagation", method );
		}

		return changed;
	}
}

File: scala/apparat/src/apparat/taas/optimization/CopyPropagation.scala
/*
 * This file is part of Apparat.
 *
 * Apparat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Apparat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Apparat. If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright (C) 2010 Joa Ebert
 * http://www.joa-ebert.com/
 *
 */
package apparat.taas.optimization

import apparat.taas.ast._
import apparat.taas.graph.TaasGraph
import annotation.tailrec

/**
 * @author Joa Ebert
 */
object CopyPropagation extends TaasOptimization {
	def name = "Copy Propagation"
	
	def optimize(context: TaasOptimizationContext) = apply(context.code.graph) match {
		case true => context.copy(modified = true)
		case false => context
	}

	def apply(graph: TaasGraph): Boolean = {
		var modified = false
		for(vertex <- graph.verticesIterator) {
			val (m, r) = propagateValuesAndExpressions(vertex.block)
			modified |= m
			vertex.block = r
		}

		modified
	}

	def propagateValuesAndExpressions(block: List[TExpr]): (Boolean, List[TExpr]) = {
		var modified = false
		var r = List.empty[TExpr]
		var values = Map.empty[Int, TValue]
		var exprs = Map.empty[Int, T3]

		@inline def usesValue(register: Int)(t: (Int, TValue)) = t._2 match {
			case TReg(index) if register == index => true
			case _ => false
		}

		@inline def usesExpr(register: Int)(t: (Int, T3)) = t._2 match {
			case T3(_, lhs: TReg, _, _) if lhs.index == register => true
			case T3(_, _, rhs: TReg, _) if rhs.index == register => true
			case _ => false
		}

		@inline def update(index: Int, valuef: Map[Int, TValue] => Map[Int, TValue], exprf: Map[Int, T3] => Map[Int, T3]): Unit = {
			values = valuef(values filterNot { usesValue(index) _ })
			exprs = exprf(exprs filterNot { usesExpr(index) _ })
		}

		@inline def map(original: TExpr): TExpr = {
			val (m, n) = copy(original, values)
			modified |= m
			r = n :: r
			n
		}

		for(op <- block) {
			op match {
				case t2 @ T2(op, rhs: TReg, result) if op == TOp_Nothing => {
					(exprs get rhs.index) match {
						case Some(expr) => {
							val n = T3(expr.op, expr.lhs, expr.rhs, result)

							modified = true
							r = n :: r

							update(result.index, _ - result.index, _ + (result.index -> n))
						}
						case None => {
							val n = map(t2)
							update(result.index, _ + (result.index -> n.asInstanceOf[T2].rhs), _ - result.index)
						}
					}
				}
				case t2 @ T2(op, rhs, result) if op == TOp_Nothing => {
					val n = map(t2)
					update(result.index, _ + (result.index -> n.asInstanceOf[T2].rhs), _ - result.index)
				}
				case t3 @ T3(op, lhs, rhs, result) => {
					val n = map(t3)
					update(result.index, _ - result.index, _ + (result.index -> n.asInstanceOf[T3]))
				}
				case tdef: TDef => {
					map(tdef)
					update(tdef.register, _ - tdef.register, _ - tdef.register)
				}
				case o => map(o)
			}
		}

		if(modified) {
			(true, r.reverse)
		} else {
			(false, block)
		}
	}

	private def copy(texpr: TExpr, available: Map[Int, TValue]): (Boolean, TExpr) = {
		texpr match {
			// x = op y -> x = op f(y)
			case t2 @ T2(op, rhs: TReg, result) => (available get rhs.index) match {
				case Some(value) => (true, T2(op, value, result))
				case None => (false, t2)
			}

			// x = y op z -> x = f(y) op f(z)
			case t3 @ T3(op, lhs: TReg, rhs: TReg, result) => (available get lhs.index) match {
				case Some(lvalue) => (available get rhs.index) match {
					case Some(rvalue) => (true, T3(op, lvalue, rvalue, result))
					case None => (true, T3(op, lvalue, rhs, result))
				}
				case None => (available get rhs.index) match {
					case Some(rvalue) => (true, T3(op, lhs, rvalue, result))
					case None => (false, t3)
				}
			}
			case t3 @ T3(op, lhs: TReg, rhs, result) => (available get lhs.index) match {
				case Some(value) => (true, T3(op, value, rhs, result))
				case None => (false, t3)
			}
			case t3 @ T3(op, lhs, rhs: TReg, result) => (available get rhs.index) match {
				case Some(value) => (true, T3(op, lhs, value, result))
				case None => (false, t3)
			}

			//if(x op y) -> if(f(x) op f(y))
			case if2 @ TIf2(op, lhs: TReg, rhs: TReg) => (available get lhs.index) match {
				case Some(lvalue) => (available get rhs.index) match {
					case Some(rvalue) => (true, TIf2(op, lvalue, rvalue))
					case None => (true, TIf2(op, lvalue, rhs))
				}
				case None => (available get rhs.index) match {
					case Some(rvalue) => (true, TIf2(op, lhs, rvalue))
					case None => (false, if2)
				}
			}
			case if2 @ TIf2(op, lhs: TReg, rhs) => (available get lhs.index) match {
				case Some(value) => (true, TIf2(op, value, rhs))
				case None => (false, if2)
			}
			case if2 @ TIf2(op, lhs, rhs: TReg) => (available get rhs.index) match {
				case Some(value) => (true, TIf2(op, lhs, value))
				case None => (false, if2)
			}

			// x = y.method(arguments) -> x = f(y).method(arguments)
			// y.method(arguments) -> f(y).method(arguments)
			case call @ TCall(t: TReg, method, arguments, result) => (available get t.index) match {
				case Some(value) => {
					(true, TCall(value, method, copy(arguments, available)._2, result))
				}
				case None => copy(arguments, available) match {
					case (true, na) => (true, TCall(t, method, na, result))
					case _ => (false, call)
				}
			}

			case tsuper @ TSuper(base: TReg, arguments) => (available get base.index) match {
				case Some(value) => (true, TSuper(value, copy(arguments, available)._2))
				case None => copy(arguments, available) match {
					case (true, na) => (true, TSuper(base, na))
					case _ => (false, tsuper)
				}
			}

			case other => (false, other)
		}
	}

	def copy(list: List[TValue], available: Map[Int, TValue]): (Boolean, List[TValue]) = {
		var m = false
		val r = list map {
			_ match {
				case reg: TReg => (available get reg.index) match {
					case Some(value) => { m = true; value }
					case None => reg
				}
				case o => o
			}
		}

		(m, r)
	}
}